clang/lib/CodeGen/CGCall.cpp

```diff
@@ -1836,6 +1837,43 @@ static bool HasStrictReturn(const CodeGenModule &Module, QualType RetTy,
          Module.getLangOpts().Sanitize.has(SanitizerKind::Return);
 }
 
+static llvm::fp::FPAccuracy convertFPAccuracy(StringRef FPAccuracyStr) {
+  return llvm::StringSwitch<llvm::fp::FPAccuracy>(FPAccuracyStr)
+      .Case("high", llvm::fp::FPAccuracy::High)
+      .Case("medium", llvm::fp::FPAccuracy::Medium)
+      .Case("low", llvm::fp::FPAccuracy::Low)
+      .Case("sycl", llvm::fp::FPAccuracy::SYCL)
+      .Case("cuda", llvm::fp::FPAccuracy::CUDA);
+}
+
+void CodeGenModule::getDefaultFunctionFPAccuracyAttributes(
+    StringRef Name, llvm::AttrBuilder &FuncAttrs, unsigned ID,
+    const llvm::Type *FuncType) {
+  // First check that a specific accuracy hasn't been requested for this
+  // function.
+  if (!getLangOpts().FPAccuracyFuncMap.empty()) {
+    auto FuncMapIt = getLangOpts().FPAccuracyFuncMap.find(Name.str());
+    if (FuncMapIt != getLangOpts().FPAccuracyFuncMap.end()) {
+      StringRef FPAccuracyVal = llvm::fp::getAccuracyForFPBuiltin(
+          ID, FuncType, convertFPAccuracy(FuncMapIt->second));
+      assert(!FPAccuracyVal.empty() && "A valid accuracy value is expected");
+      FuncAttrs.addAttribute("fpbuiltin-max-error=", FPAccuracyVal);
+    }
+  }
+  if (!getLangOpts().FPAccuracyVal.empty()) {
+    StringRef FPAccuracyVal = llvm::fp::getAccuracyForFPBuiltin(
+        ID, FuncType, convertFPAccuracy(getLangOpts().FPAccuracyVal));
+    assert(!FPAccuracyVal.empty() && "A valid accuracy value is expected");
+    if (FuncAttrs.attrs().size() == 0)
+      // No specific accuracy has been requested for this function.
+      // Add the general accuracy value to the attribute list.
+      FuncAttrs.addAttribute("fpbuiltin-max-error=", FPAccuracyVal);
+    else
+      assert(FuncAttrs.attrs().size() == 1 &&
+             "Expected an attribute has already been mapped to the function");
```

Comment:
I am not able to understand this despite multiple readings.
Can you add some comments with an example perhaps to show what each if condition does, as well as the size checks here?

Reply:
Added comment. Let me know if this is satisfactory.